### ISSUE TITLE
JDBC config-based exception identification

### ISSUE THREAD

**Issue by aguibert** [OpenLiberty/open-liberty#14713]
REF: http://www.ibm.com/developerworks/rfe/execute?use_case=viewChangeRequest&CR_ID=93960
UFO doc: https://ibm.box.com/s/samdsp3h4qxirsbeayokxtargu06f34c

## User Story:
As a user of a JDBC driver that throws non-standard JDBC exceptions, I want to be able to configure which SQL Error Codes and SQL States correspond to certain error categories so my connection pool behaves more effeciently.

## Detail
New config element `<identifyException>` (child of `<dataSource>`)
Attributes:
- `errorCode`: (int) The vendor error code of the SQLException to identify
- `sqlState`: (String) The SQL State of the SQLException to identify
- `as`: (String) The state to classify the errorCode and/or sqlState as. Can be one of: StaleConnection, StaleStatement, Unsupported, None, or a fully qualified class name

Example configuration:
```xml
<featureManager>
  <feature>jdbc-4.2</feature>
</featureManager>

<dataSource jndiName="jdbc/myDB">
  <jdbcDriver ... />
  <properties ... />
  <identifyException errorCode="1234" as="StaleConnection"/>
  <identifyException sqlState="E001"  as="StaleStatement"/>
  <identifyException sqlState="E002"  as="Unsupported"/>
  <identifyException sqlState="E003" errorCode="5678" as="None"/>
  <identifyException sqlState="E004"  as="com.foo.MySQLException"/>
</dataSource>
```

**Comment by jwalcorn** [comment]
We hit this in a modernization POC with a customer. Their code was using Spring’s WebSphereDataSourceAdapter, which uses some APIs from tWAS that aren’t in Liberty. If this gets that Spring class working in Liberty, that would simply efforts to port legacy on-premises tWAS apps to a Liberty container, such as running on OpenShift in the cloud.

**Comment by samwatibm** [comment]
RFE's have moved: https://cloud-platform.ideas.aha.io/ideas/TWAS-I-224

**Comment by njr-11** [comment]
@donbourne - for the focalApproved:serviceability approval,

Serviceability Approval Comment:

WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? Have these issues been addressed in the implementation?
No, this information belongs in the Skills Transfer and will be put there instead.
Yes, we are covering serviceability paths/issues in the implementation.

Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).
a) What problem paths were tested and demonstrated?
Configuring a custom SQLException class name that is misspelled or doesn't exist.
And other minor messages to the user.
b) Who did you demo to?
Greg Watts, Kyle Aure, Nathan Mittlestat
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?
Mostly. A couple of improvements to error paths were requested, which I made as a result.

SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
No one. It was determined that SVT is unnecessary due to exhaustive coverage in FAT, which we wrote ourselves.
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?
N/A.

Which L2 / L3 queues will handle PMRs for this feature? Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it. Ask Don Bourne if you need links or more info.
WAS: L2 CET
WAS: L3 JDBC

**Comment by njr-11** [comment]
@skasund here is the skills transfer doc,
https://ibm.box.com/s/dh6n0hocb4sw6ob9lgm3f22qo1kahlh2
Please add the `focalApproved:ste` label.

**Comment by donbourne** [comment]
@njr-11 re your comment... 
> No, this information belongs in the Skills Transfer and will be put there instead.

...The rationale for asking for it in the UFO is to show people what you're thinking will be problems customers may encounter and give them a chance to comment if they can foresee other issues or other ways of handling the problems.

**Comment by njr-11** [comment]
> ...The rationale for asking for it in the UFO is to show people what you're thinking will be problems customers may encounter and give them a chance to comment if they can foresee other issues or other ways of handling the problems.

Thanks for pointing out the reason behind it. I'll try to remember to include this information in the future Upcoming Feature Overview documents.

**Comment by chirp1** [comment]
From Nathan, no ID work is required for this feature. Documentation is auto-generated. Usage is covered under another feature. Approved
